survivalstan/formulas.py

```diff
@@ -0,0 +1,315 @@
+import pandas as pd
+import patsy
+import sys
+import numpy as np
+import re
+import logging
+logger = logging.getLogger(__name__)
+
+def _prep_timepoint_dataframe(df,
+                              timepoint_end_col,
+                              timepoint_id_col = None
+                              ):
+    """ Helper function to take a set of timepoints
+        in observation-level dataframe & return
+        formatted timepoint_id, end_time, duration
+
+        Returns
+        ---------
+        pandas dataframe with one record per timepoint_id
+            where timepoint_id is the index
+            sorted on the index, increasing
+
+    """
+    time_df = df.copy()
+    time_df.sort_values(timepoint_end_col, inplace=True)
+    if not(timepoint_id_col):
+        timepoint_id_col = 'timepoint_id'
+        time_df[timepoint_id_col] = time_df[timepoint_end_col].astype('category').cat.codes + 1
+    time_df.dropna(how='any', subset=[timepoint_id_col, timepoint_end_col], inplace=True)
+    time_df = time_df.loc[:,[timepoint_id_col, timepoint_end_col]].drop_duplicates()
+    time_df[timepoint_end_col] = time_df[timepoint_end_col].astype(np.float32)
+    time_df.set_index(timepoint_id_col, inplace=True, drop=True)
+    time_df.sort_index(inplace=True)
+    t_durs = time_df.diff(periods=1)
+    t_durs.rename(columns = {timepoint_end_col: 't_dur'}, inplace=True)
+    time_df = time_df.join(t_durs)
+    if len(time_df.index)>1:
+        time_df.fillna(inplace=True, value=time_df.loc[1, timepoint_end_col])
+    return(time_df)
+
+
+class Id(object):
+    def __init__(self, desc='id'):
+        self.values = []
+        self.desc = desc
+
+    def memorize_chunk(self, x):
+        self.values.extend(np.unique(x))
+
+    def memorize_finish(self):
+        self.ids = np.arange(len(self.values))+1
+        self.lookup = dict(zip(self.values, self.ids))
+
+    def transform(self, x):
+        if patsy.util.have_pandas and isinstance(x, pd.Series):
+            d = pd.Series([self.lookup[val] for val in x]).astype(int)
+            d.index = x.index
+            return(d)
+        else:
+            return np.array([self.lookup[val] for val in x])
+
+    def len(self):
+        return len(self.ids)
+
+    def decode_df(self):
+        df = pd.DataFrame({'id': self.ids, 'value': self.values})
+        df.dropna(inplace=True)
+        return df
+
+as_id = patsy.stateful_transform(Id)
+
+class SurvData(pd.DataFrame):
+    ''' patsy.DesignMatrix representing survival data output '''
+    survival_type = 'wide'
```

Comment:
I like the first idea of using `SurvData` as a base class & having both `WideSurvData` & `LongSurvData` extending it. In general this is meant to be a starting point, since I will likely be adding new data types for (e.g.) semi-competing risks models, etc.